packages/syrup/src/decode.js

```diff
@@ -0,0 +1,345 @@
+// @ts-check
+
+import { compareByteArrays } from './compare.js';
+
+const MINUS = '-'.charCodeAt(0);
+const ZERO = '0'.charCodeAt(0);
+const ONE = '1'.charCodeAt(0);
+const NINE = '9'.charCodeAt(0);
+const LIST_START = '['.charCodeAt(0);
+const LIST_END = ']'.charCodeAt(0);
+const DICT_START = '{'.charCodeAt(0);
+const DICT_END = '}'.charCodeAt(0);
+// const SET_START = '#'.charCodeAt(0);
+// const SET_END = '$'.charCodeAt(0);
+const BYTES_START = ':'.charCodeAt(0);
+const STRING_START = '"'.charCodeAt(0);
+const SYMBOL_START = "'".charCodeAt(0);
+// const RECORD_START = '<'.charCodeAt(0);
+// const RECORD_END = '>'.charCodeAt(0);
+const INT_START = 'i'.charCodeAt(0);
+const INT_END = 'e'.charCodeAt(0);
+const NULL = 'n'.charCodeAt(0);
+const UNDEFINED = 'u'.charCodeAt(0);
+const TRUE = 't'.charCodeAt(0);
+const FALSE = 'f'.charCodeAt(0);
+const SINGLE = 'F'.charCodeAt(0);
+const DOUBLE = 'D'.charCodeAt(0);
+
+const CANONICAL_NAN_32 = new Uint8Array(4);
+new DataView(CANONICAL_NAN_32.buffer).setFloat32(0, NaN);
+
+const CANONICAL_NAN_64 = new Uint8Array(8);
+new DataView(CANONICAL_NAN_64.buffer).setFloat64(0, NaN);
+
+const textDecoder = new TextDecoder();
+
+const scratch = new ArrayBuffer(8);
+const scratchBytes = new Uint8Array(scratch);
+const scratchData = new DataView(scratch);
```

Comment:
TypeScript makes it hard for this to be anything else without having to create a modified version of `compareByteArrays`.

Reply:
Seems ironic for our type checker to force us to use patterns that are less safe and even less statically checkable. If that's the motivation, I'd prefer to keep the modules checkably pure and disable the particular typescript complaints that are getting in the way.

Reply:
Added an `isCanonicalNaN64` function that hard codes the NaN that Python and Node.js agree upon. We have the option of removing forced canonicalization of -0 and NaN if that’s not necessary for three-party-handoff. I simply don’t know the ramifications.

Reply:
That's fine. But to be clear, I was complaining about the `scratch*` stuff since it is mutated at runtime. The NaN array didn't really bother me because after initialization it is never subsequently mutated. I was just concerned that it be canonical.